feat: experimental WebDriver BiDi support with Firefox (#11412)
puppeteer/puppeteer: docs/webdriver-bidi.md

# Experimental WebDriver BiDi Support

[WebDriver BiDi](https://w3c.github.io/webdriver-bidi/) is a new cross-browser

> automation protocol that extends WebDriver with events. Puppeteer aims at

Reviewer: ```suggestion
automation protocol that extends WebDriver with events. Here are the resources if you want to learn more about WebDriver BiDi:

- [WebDriver BiDi - The future of cross-browser automation](https://developer.chrome.com/articles/webdriver-bidi/)
- [WebDriver BiDi: 2023 status update](https://developer.chrome.com/blog/webdriver-bidi-2023/)

## Automate with Chrome and Firefox

```

> supporting WebDriver BiDi whenever possible and currently has experimental

Reviewer: @mathiasbynens on the wording - Based on the current % features supported, is "Experimental" or "Beta" a more suitable word?

From the [definition in wikipedia](https://en.wikipedia.org/wiki/Software_release_life_cycle#:~:text=finally%20gold%20status.-,Beta,-%5Bedit%5D) - 
A beta phase generally begins when the software is **feature-complete** but likely to contain several known or unknown bugs.

Reviewer 2: “Experimental” or “Alpha” seems appropriate. One potential issue is that it’s technically been “experimentally implemented” for a few months already, but what’s new now is that we’ve reached a significant milestone of being able to deprecate the CDP-based Firefox support. I’m not sure if we need a different term for that though — maybe just sticking to “experimental” in combination with this announcement is simplest.

Reviewer: SGTM. Let's stick to "experimental".

> support for automating Firefox and Chrome over WebDriver BiDi with Firefox

Reviewer: ```suggestion
Puppeteer aims at support for automating Firefox and Chrome over WebDriver BiDi. Firefox
```

# Experimental WebDriver BiDi Support

[WebDriver BiDi](https://w3c.github.io/webdriver-bidi/) is a new cross-browser
automation protocol that adds browser-driven events to WebDriver. Here are the
resources if you want to learn more about WebDriver BiDi:

- [WebDriver BiDi - The future of cross-browser automation](https://developer.chrome.com/articles/webdriver-bidi/)
- [WebDriver BiDi: 2023 status update](https://developer.chrome.com/blog/webdriver-bidi-2023/)

## Automate with Chrome and Firefox

Firefox support has almost reaching feature parity with the previous CDP-based
implementation. To see which features are fully supported with WebDriver BiDi we
used the [Puppeteer test suite](https://puppeteer.github.io/ispuppeteerwebdriverbidiready/). Currently,
we still have [60](https://puppeteer.github.io/ispuppeteerwebdriverbidiready/firefox-delta.json) tests
that are failing with Firefox and WebDriver BiDi compared to the previous CDP

> implementation in Firefox but we also have [82](https://puppeteer.github.io/ispuppeteerwebdriverbidiready/firefox-delta.json)

Reviewer 3: How about we add `less than` `more than` to the numbers that way we don't have to feel the need to update them all the time?

  product: 'firefox',
  protocol: 'webDriverBiDi',
});
const page = await browser.newPage();
...
await browser.close();
```

Example of launching Chrome with WebDriver BiDi:

```ts
import puppeteer from 'puppeteer';

const browser = await puppeteer.launch({
  product: 'chrome',
  protocol: 'webDriverBiDi',
});
const page = await browser.newPage();
...
await browser.close();
```

## Puppeteer features supported over WebDriver BiDi

- Browser and page automation

  - Puppeteer.launch
  - Browser.close
  - Page.bringToFront
  - Page.goto
  - Page.reload (except for `ignoreCache` parameter)
  - Page.setViewport (`width`, `height`, `deviceScaleFactor` only)

- [Script evaluation](https://pptr.dev/guides/evaluate-javascript):

  - Page.evaluate
  - Page.exposeFunction
  - JSHandle.evaluate
  - JSHandle.evaluateHandle

- [Selectors](https://pptr.dev/guides/query-selectors) and [locators](https://pptr.dev/guides/locators) except for ARIA:

  - Page.$ (ARIA selectors only supported in Chrome)
  - Page.$$ (ARIA selectors only supported in Chrome)
  - Page.$eval (ARIA selectors only supported in Chrome)
  - Page.$$eval (ARIA selectors only supported in Chrome)

>   - Frame.goto() (except `referer` and `referrerPolicy`)

Reviewer 3: This should be part of `Browser and page automation`. And should affect the `Page.go`

  - Keyboard.press
  - Keyboard.sendCharacter
  - Keyboard.type
  - Keyboard.up
  - Mouse events (except for dedicated drag'n'drop API methods)
  - TouchScreen.\*
  - Page.tap

- JavaScript dialog interception

  - page.on('dialog')
  - Dialog.\*

- Screenshots (not all parameters are supported)

  - Page.screenshot (supported parameters `clip`, `encoding`, `fullPage`)

- PDF generation (not all parameters are supported)

  - Page.pdf (only `format`, `height`, `landscape`, `margin`, `pageRanges`, `printBackground`, `scale`, `width` are supported)
  - Page.createPDFStream (only `format`, `height`, `landscape`, `margin`, `pageRanges`, `printBackground`, `scale`, `width` are supported)

## Puppeteer features not yet supported over WebDriver BiDi

- [Request interception](https://pptr.dev/guides/request-interception)

  - HTTPRequest.client()
  - HTTPRequest.continueRequestOverrides()
  - HTTPRequest.responseForRequest()
  - HTTPRequest.abortErrorReason()
  - HTTPRequest.interceptResolutionState()
  - HTTPRequest.isInterceptResolutionHandled()
  - HTTPRequest.finalizeInterceptions()
  - HTTPRequest.failure()
  - HTTPRequest.continue()
  - HTTPRequest.respond()
  - HTTPRequest.abort()

>   - Page.setRequestInterception()

Reviewer 3: Let's make it the first entry of the Request interception list.

Author: I have sorted alphabetically

- Permissions

  - BrowserContext.overridePermissions()
  - BrowserContext.clearPermissionOverrides()

> - Various emulations (most are only supported with Chrome)

Reviewer 3: ```suggestion
- Various emulations (most are supported with Chrome)
```

Reviewer 3: This should be a little bit better as we tell - These features are unsupported by BiDi and here we inverse it to Most emulations are supported by BiDi for Chrome.
